전체 상품 조회 페이지 기능

riassuc opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 로그인 버튼 활성화 기능 구현(예시)
- 전체 상품 조회 페이지 기능 구현
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- id '@' 포함, pw 5 글자 이상 입력 시 버튼 활성화(예시)
- 필터링이나 정렬 없이 db순으로 전체 상품 조회
- 요일별 / 맛별 필터링 기능 구현
- 신메뉴, 인기메뉴, 가격, 만족도순으로 정렬 기능 구현
- 페이지네이션 기능 구현
<img width="835" alt="스크린샷 2021-06-14 오전 9 41 39" src="https://user-images.githubusercontent.com/76906649/121828033-46b32280-ccf9-11eb-83bd-a16922bff471.png">


## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 함수 로직이 너무 긴 것 같습니다. 좀 더 효율적인 방법이 없을지 궁금합니다.(예시) 
- 일단 아직 전체 리스트 조회만 가능하고 필터링, 정렬, 페이지네이션, restfull한 경로, 에러핸들링 미구현된 상태입니다.
- 쿼리파라미터를 사용할때 
   `~경로~/?sort=price&taste=`
   `~경로~/?sort=price`
   이렇게 taste= 처럼 값이 없는 항목은 안보이는게 rest api에 더 맞다고 생각하는데 실제 웹사이트들을 보면 값이 없는데도
   주소가 있는 경우가 많아서 뭐가 restful 한건지 궁금합니다

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

riassuc committed:
Add: 전체상품 조회페이지 구현

riassuc committed:
Modify: 불필요한 모듈 삭제

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,33 @@
+from products.models import Product
+
+class ProductList(View):
+
+    def get(self, request):
+
```
불필요한 공백은 없애주세요.

riassuc replied:
넵!!

redsaumon commented on products/views.py:
```diff
@@ -1,3 +1,33 @@
+
+    def get(self, request):
+
+        products = Product.objects.all()
+        
+        result = []
+
+        for product in products:
+
+            images     = product.productimage_set.all()
+            image_list = [image.image_url for image in images]
+            tastes     = product.tasteproduct_set.all()
+            taste      = [taste.taste.taste for taste in tastes]
+
+            product_info = {
+                'title'     : product.title,
+                'sub_title' : product.sub_title,
+                'price'     : product.price,
+                'calorie'   : product.calorie,
+                'gram'      : product.gram,
+                'taste'     : taste,
+                'images'    : image_list
+            }
+
+            result.append(product_info)
```
좋습니다! product_info를 만드는 부분도 list_comprehension을 사용해서 작성해주세요. 간단한 로직일 경우 일반 for loop보다 속도가 2배 가량 빠릅니다.

riassuc replied:
넵! 수정했습니다.

riassuc committed:
Modify: 리팩토링

riassuc committed:
Modify: 버그 수정
- shell에서 이상이 없어서 push 했는데 알고보니 잘못된 부분이 있었습니다.
- 앞으로는 이상없이 작동하는것 확인하고 push 하겠습니다.

riassuc committed:
Fix: 충돌 해결

riassuc committed:
Fix: 충돌 해결

riassuc committed:
Fix: 진짜해결

riassuc committed:
Add: 필터, 정렬, 페이지네이션 구현
- 요일, 맛으로 필터링 가능
- 신메뉴, 인기메뉴, 가격, 만족도순으로 정렬 가능
- 페이지네이션 구현

riassuc committed:
Modify: 코드 컨벤션에 맞춰 수정

redsaumon commented on products/urls.py:
```diff
@@ -1,7 +1,7 @@
-
+from products.views import ProductList, ProductDetailView
 urlpatterns = [
+    path('/', ProductList.as_view()),
```
이렇게 되면 엔드포인트가 products/ 와 같은 형태인데 RESTful하지 못합니다. /는 없애주세요!

riassuc replied:
넵!!!

redsaumon commented on products/views.py:
```diff
@@ -28,4 +30,43 @@ def get(self, request, product_title):
+            rating=Avg('review__star_rating'), day=F('dayproduct__day'), taste=F('tasteproduct__taste')).filter(
+                Q(day=day)|Q(taste=taste)).order_by(sort)
+        
+        objects_count = products.count()
```
이정도 코드는 변수화하지 않아도 될 거 같습니다.

riassuc replied:
이 부분은 결과값에
페이지네이션 되기 전 products 객체의 전체 개수,
페이지네이션 된 후 현재 페이지에 보이는 개수를 각각 넣고 싶어서 의도한 부분입니다!

redsaumon commented on products/views.py:
```diff
@@ -28,4 +30,43 @@ def get(self, request, product_title):
+        day      = request.GET.get('day')
+        taste    = request.GET.get('taste')
+        page     = int(request.GET.get('page', 1))
+        per_page = int(request.GET.get('per_page', 20))
```
20이라는 per_page를 상수화 시켜주세요.

riassuc replied:
넵!!!!

redsaumon commented on products/views.py:
```diff
@@ -28,4 +30,43 @@ def get(self, request, product_title):
+        
+        objects_count = products.count()
+        
+        if page > 0 and per_page > 0:
```
이 if문을 타지 않으면 전체 상품이 한번에 가게 의도하신 걸까요? 혹시 모를 예외 상황을 피하기 위한 코드라면, 그에 해당하는 return 값도 작성해주세요!

riassuc replied:
이 부분은  표현할수 있는 페이지 범위를 넘어가면 페이지를 찾을 수 없다는 에러가 나오게 수정 해봤습니다.

redsaumon commented on products/views.py:
```diff
@@ -28,4 +30,43 @@ def get(self, request, product_title):
+            'calorie'   : product.calorie,
+            'gram'      : product.gram,
+            'taste'     : [taste.taste.name for taste in product.tasteproduct_set.all()],
+            'images'    : '' if product.productimage_set.all().first() == None else product.productimage_set.all().first().image_url
```
```suggestion
            'images'    : None if product.productimage_set.all().first() is None else product.productimage_set.all().first().image_url
```

riassuc replied:
넵!!!!!

redsaumon commented on products/views.py:
```diff
@@ -28,4 +30,43 @@ def get(self, request, product_title):
+            ended    = started + per_page
+            products = products[started:ended]
+
+        result = {
+            'count'    : objects_count,
+            'page'     : page,
+            'per_page' : per_page,
+            'pages'    : math.ceil(objects_count/per_page),
+            'elements' : products.count()
+        }
+        
+        result['result'] = [{
+            'title'     : product.title,
+            'sub_title' : product.sub_title,
+            'price'     : product.price,
+            'calorie'   : product.calorie,
+            'gram'      : product.gram,
+            'taste'     : [taste.taste.name for taste in product.tasteproduct_set.all()],
+            'images'    : '' if product.productimage_set.all().first() == None else product.productimage_set.all().first().image_url
+        } for product in products]
```
하나의 result로 합쳐주세요!

riassuc replied:
넵 위의 정보부분 리스트의 맨마지막에 추가하게 수정했습니다.

redsaumon commented on products/views.py:
```diff
@@ -28,4 +30,43 @@ def get(self, request, product_title):
+            'images'    : '' if product.productimage_set.all().first() == None else product.productimage_set.all().first().image_url
+        } for product in products]
+
+        return JsonResponse(result, status=200)
```
보내는 data의 key 값도 지정해주세요.

riassuc replied:
넵 result를 딕셔너리에서 다시 리스트로 바꾸면서 추가했습니다.

riassuc replied:
<img width="1536" alt="스크린샷 2021-06-14 오후 9 17 23" src="https://user-images.githubusercontent.com/76906649/121891011-f0c09800-cd55-11eb-91a0-2da471a3feb5.png">

riassuc committed:
Modify: 반환값에 id추가

riassuc committed:
Fix: 충돌 해결

riassuc committed:
Modify: 필터링 잘못되던 부분 수정

redsaumon commented:
준님 수정사항 반영이 안된 거 같아요! 확인 한번 해주세요!

riassuc commented:
네 하다가 밥먹고 왔습니다 얼릉 해보겠습니다!

riassuc committed:
Fix: 충돌

riassuc committed:
Modify: 리뷰수정사항 반영

redsaumon commented on products/views.py:
```diff
@@ -57,4 +59,51 @@ def get(self, request, product_title):
+            'price'     : product.price,
+            'calorie'   : product.calorie,
+            'gram'      : product.gram,
+            'taste'     : None if product.tasteproduct_set.first() == None else product.tasteproduct_set.first().taste.name,
+            'images'    : None if product.productimage_set.first() == None else product.productimage_set.first().image_url
```
None을 비교할 때는 보통 is를 사용합니다. == 도 가능하지만 is보다 성능이 좋지 않습니다.

riassuc replied:
넵!!

riassuc committed:
Modify: 수정사항 반영
-per_page 상수화
-ProductList 메인페이지에서 재사용 할수 있게 수정

riassuc committed:
Fix: 충돌해결

riassuc commented:
메인페이지에서 요청을 할때는 per_page라는 쿼리파라미터를 사용해서 6개만 보낼 수 있게 했는데,
상수로 사용하시라고 한 부분때문에 뭔가 레스트풀하지 못한것 같습니다. 
인터넷에서 페이지네이션 검색해보면 그냥 변수로 많이들 쓰고, 프론트에서 요청받을때 리밋, 오프셋 두개 다 같이 받는 경우가 많이 있는것 같은데 연우님께서 PER_PAGE를 상수로 사용하라고 하신 이유가 있을까요??

riassuc committed:
Modify: 반환되는 데이터 추가

redsaumon commented:
수고하셨습니다~
